File: app/src/main/java/com/kaltura/kflow/presentation/player/PlaybackControlsView.java
package com.kaltura.kflow.presentation.player;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import com.google.android.exoplayer2.C;
import com.kaltura.client.types.Asset;
import com.kaltura.kflow.R;
import com.kaltura.playkit.PKLog;
import com.kaltura.playkit.Player;
import com.kaltura.playkit.PlayerState;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Formatter;
import java.util.Locale;

/**
 * Created by anton.afanasiev on 07/11/2016.
 */

public class PlaybackControlsView extends LinearLayout implements View.OnClickListener, SeekBar.OnSeekBarChangeListener {

    private static final PKLog log = PKLog.get("PlaybackControlsView");
    private static final int PROGRESS_BAR_MAX = 100;

    private Player player;
    private PlayerState playerState;
    private Asset asset;

    private Formatter formatter;
    private StringBuilder formatBuilder;

    private SeekBar seekBar;
    private TextView tvCurTime, tvTime;
    private ImageButton btnPlay, btnPause, btnStartOver;

    private boolean dragging = false;

    private Runnable updateProgressAction = this::updateProgress;

    public PlaybackControlsView(Context context) {
        this(context, null);
    }

    public PlaybackControlsView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PlaybackControlsView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(context).inflate(R.layout.view_player_control, this);
        formatBuilder = new StringBuilder();
        formatter = new Formatter(formatBuilder, Locale.getDefault());
        initPlaybackControls();
    }

    private void initPlaybackControls() {

        btnPlay = findViewById(R.id.play);
        btnPause = findViewById(R.id.pause);
        btnStartOver = findViewById(R.id.startover);

        btnPlay.setOnClickListener(this);
        btnPause.setOnClickListener(this);

        seekBar = findViewById(R.id.mediacontroller_progress);
        seekBar.setOnSeekBarChangeListener(this);

        tvCurTime = findViewById(R.id.time_current);
        tvTime = findViewById(R.id.time);
    }

    private void updateProgress() {
        if (asset == null) updateVodProgress();
        else updateLiveProgress();
    }

    private void updateLiveProgress() {
        long startDate = asset.getStartDate();
        long endDate = asset.getEndDate();
        long durationMs = endDate - startDate;
        long currentProgress = new Date().getTime() / 1000 - startDate;

        SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm", Locale.getDefault());
        String startTime = dateFormat.format(new Date(startDate));
        String endTime = dateFormat.format(new Date(endDate));

        tvCurTime.setText(startTime);
        tvTime.setText(endTime);

        seekBar.setProgress((int) ((float) PROGRESS_BAR_MAX * currentProgress / durationMs));
    }

    private void updateVodProgress() {
        long duration = C.TIME_UNSET;
        long position = C.POSITION_UNSET;
        long bufferedPosition = 0;
        if (player != null) {
            duration = player.getDuration();
            position = player.getCurrentPosition();
            bufferedPosition = player.getBufferedPosition();
        }

        if (duration != C.TIME_UNSET) {
            log.d("updateProgress Set Duration:" + duration);
            tvTime.setText(stringForTime(duration));
        }

        if (!dragging && position != C.POSITION_UNSET && duration != C.TIME_UNSET) {
            log.d("updateProgress Set Position:" + position);
            tvCurTime.setText(stringForTime(position));
            seekBar.setProgress(progressBarValue(position));
        }

        seekBar.setSecondaryProgress(progressBarValue(bufferedPosition));
        // Remove scheduled updates.
        removeCallbacks(updateProgressAction);
        // Schedule an update if necessary.
        if (playerState != PlayerState.IDLE) {
            long delayMs = 1000;
            postDelayed(updateProgressAction, delayMs);
        }
    }

    private int progressBarValue(long position) {
        int progressValue = 0;
        if (player != null) {
            long duration = player.getDuration();
            if (duration > 0) {
                progressValue = (int) ((position * PROGRESS_BAR_MAX) / duration);
            }
        }

        return progressValue;
    }

    private long positionValue(int progress) {
        long positionValue = 0;
        if (player != null) {
            long duration = player.getDuration();
            positionValue = (duration * progress) / PROGRESS_BAR_MAX;
        }

        return positionValue;
    }

    private String stringForTime(long timeMs) {

        long totalSeconds = (timeMs + 500) / 1000;
        long seconds = totalSeconds % 60;
        long minutes = (totalSeconds / 60) % 60;
        long hours = totalSeconds / 3600;
        formatBuilder.setLength(0);
        return hours > 0 ? formatter.format("%d:%02d:%02d", hours, minutes, seconds).toString()
                : formatter.format("%02d:%02d", minutes, seconds).toString();
    }

    public void setPlayer(Player player) {
        this.player = player;
    }

    public void setPlayerState(PlayerState playerState) {
        this.playerState = playerState;
        updateProgress();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.play:
                if (player != null) {
                    player.play();
                }
                break;
            case R.id.pause:
                player.pause();
                break;
        }
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (fromUser) {
            tvCurTime.setText(stringForTime(positionValue(progress)));
        }
    }


    public void onStartTrackingTouch(SeekBar seekBar) {
        dragging = true;
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        dragging = false;
        player.seekTo(positionValue(seekBar.getProgress()));
    }

    public void setOnStartOverClickListener(View.OnClickListener listener) {
        btnStartOver.setOnClickListener(listener);
    }

    public void release() {
        removeCallbacks(updateProgressAction);
    }

    public void resume() {
        updateProgress();
    }

    public void disableControllersForLive() {
        btnPause.setEnabled(false);
        btnPlay.setEnabled(false);

        btnPause.getDrawable().mutate().setTint(getResources().getColor(android.R.color.darker_gray));
        btnPlay.getDrawable().mutate().setTint(getResources().getColor(android.R.color.darker_gray));

        seekBar.setOnTouchListener((v, event) -> true);
    }

    public void setAsset(Asset asset) {
        this.asset = asset;
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/player/PlaybackControlsView.kt
package com.kaltura.kflow.presentation.player

import android.content.Context
import android.util.AttributeSet
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import com.kaltura.android.exoplayer2.C
import com.kaltura.client.types.Asset
import com.kaltura.kflow.R
import com.kaltura.kflow.presentation.extension.getColor
import com.kaltura.kflow.presentation.extension.inflate
import com.kaltura.playkit.PKLog
import com.kaltura.playkit.PlayerState
import com.kaltura.tvplayer.KalturaPlayer
import kotlinx.android.synthetic.main.view_player_control.view.*
import java.text.SimpleDateFormat
import java.util.*
import android.provider.Settings.System.DATE_FORMAT
import android.util.Log
import java.text.DateFormat


class PlaybackControlsView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private val log = PKLog.get("PlaybackControlsView")
    private val PROGRESS_BAR_MAX = 100

    var player: KalturaPlayer? = null
    var asset: Asset? = null
    var isLive = false
    var onPlayerLiveSeek: (positionMs: Long) -> Unit = {}
    private var pausedPosition = 0L
    private var playerState: PlayerState = PlayerState.IDLE
    private val formatBuilder = StringBuilder()
    private val formatter = Formatter(formatBuilder, Locale.getDefault())
    private var dragging = false
    private val updateProgressAction = Runnable { updateProgress() }

    init {
        inflate(R.layout.view_player_control, true)
        play.setOnClickListener {
            if (isLive) {
                onPlayerLiveSeek(pausedPosition)
            } else {
                player?.play()
            }
        }
        pause.setOnClickListener {
            if (isLive) {
                pausedPosition =
                    Calendar.getInstance(TimeZone.getTimeZone("UTC")).timeInMillis / 1000 - (asset?.startDate
                        ?: 0L)
                Log.e("pause", "pausedPosition = $pausedPosition")
            }
            player?.pause()
        }
        mediacontrollerProgress.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    timeCurrent.text = stringForTime(positionValue(progress))
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                dragging = true
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                dragging = false
                player?.seekTo(positionValue(seekBar.progress))
            }

        })
    }

    private fun updateProgress() {
        if (isLive) updateLiveProgress() else updateVodProgress()
    }

    private fun updateLiveProgress() {
        val startDate = asset?.startDate ?: 0L
        val endDate = asset?.endDate ?: 0L
        val durationMs = endDate - startDate

        val startDayCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        startDayCalendar.timeInMillis = startDate * 1000
        val endDayCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        endDayCalendar.timeInMillis = endDate * 1000

        val currentProgress =
            Calendar.getInstance(TimeZone.getTimeZone("UTC")).timeInMillis / 1000 - startDate
        val dateFormat = SimpleDateFormat("HH:mm", Locale.US)
        val startTime = dateFormat.format(startDayCalendar.time)
        val endTime = dateFormat.format(endDayCalendar.time)
        timeCurrent.text = startTime
        time.text = endTime
        mediacontrollerProgress.progress = (PROGRESS_BAR_MAX * currentProgress / durationMs).toInt()
    }

    private fun updateVodProgress() {
        val duration = player?.duration ?: C.TIME_UNSET
        val position = player?.currentPosition ?: C.POSITION_UNSET.toLong()
        val bufferedPosition = player?.bufferedPosition ?: 0L

        if (duration != C.TIME_UNSET) {
            log.d("updateProgress Set Duration:$duration")
            time.text = stringForTime(duration)
        }
        if (!dragging && position != C.POSITION_UNSET.toLong() && duration != C.TIME_UNSET) {
            log.d("updateProgress Set Position:$position")
            timeCurrent.text = stringForTime(position)
            mediacontrollerProgress.progress = progressBarValue(position)
        }
        mediacontrollerProgress.secondaryProgress = progressBarValue(bufferedPosition)
        // Remove scheduled updates.
        removeCallbacks(updateProgressAction)
        // Schedule an update if necessary.
        if (playerState != PlayerState.IDLE) {
            postDelayed(updateProgressAction, 1000L)
        }
    }

    private fun progressBarValue(position: Long): Int {
        var progressValue = 0
        if (player != null) {
            val duration = player!!.duration
            if (duration > 0) {
                progressValue = (position * PROGRESS_BAR_MAX / duration).toInt()
            }
        }
        return progressValue
    }

    private fun positionValue(progress: Int): Long {
        var positionValue: Long = 0
        if (player != null) {
            val duration = player!!.duration
            positionValue = duration * progress / PROGRESS_BAR_MAX
        }
        return positionValue
    }

    private fun stringForTime(timeMs: Long): String {
        val totalSeconds = (timeMs + 500) / 1000
        val seconds = totalSeconds % 60
        val minutes = totalSeconds / 60 % 60
        val hours = totalSeconds / 3600
        formatBuilder.setLength(0)
        return if (hours > 0) formatter.format("%d:%02d:%02d", hours, minutes, seconds)
            .toString() else formatter.format("%02d:%02d", minutes, seconds).toString()
    }

    fun setPlayerState(playerState: PlayerState) {
        this.playerState = playerState
        updateProgress()
    }

    fun setOnStartOverClickListener(listener: OnClickListener?) {
        startover.setOnClickListener(listener)
    }

    fun release() = removeCallbacks(updateProgressAction)

    fun resume() = updateProgress()

    fun disableControllersForLive() {
        pause.isEnabled = false
        play.isEnabled = false
        pause.drawable.mutate().setTint(getColor(android.R.color.darker_gray))
        play.drawable.mutate().setTint(getColor(android.R.color.darker_gray))
        mediacontrollerProgress.setOnTouchListener { _, _ -> true }
    }
}